aten/src/ATen/core/TensorOptions.h

```diff
@@ -132,54 +130,54 @@ struct CAFFE2_API TensorOptions {
   }
 
   /// Returns the device of the `TensorOptions`.
-  const Device& device() const noexcept {
-    return device_;
-  }
+  Device device() const noexcept;
+
+  optional<Device> device_opt() const noexcept { return device_; }
 
   /// Returns the device index of the `TensorOptions`.
   int32_t device_index() const noexcept {
-    return device_.index();
+    return device().index();
   }
 
   /// Returns the dtype of the `TensorOptions`.
-  ScalarType dtype() const noexcept {
-    return dtype_;
-  }
+  ScalarType dtype() const noexcept;
+
+  optional<ScalarType> dtype_opt() const noexcept { return dtype_; }
 
   /// Returns the layout of the `TensorOptions`.
-  Layout layout() const noexcept {
-    return layout_;
-  }
+  Layout layout() const noexcept;
```

Comment:
I remember when I wrote this it was super high pri to have everything inline. People even pushed back on me putting these methods in the .cpp file to fix the nvcc bug you're fixing. Just so that you're aware of this. Also CC @colesbury